File: src/algorithms/arrandstring/ReverseWordsInAString.java
package algorithms.detail; 

import java.util.*; 

public class ReverseWordsInAString {
    /* Function: Reverse words and output as string without trailing spaces. 
     * Solution: Use split to devide the words, skip "" and append to a string buffer. 
     */ 
    
    public String reverseWords(String s) {
        if (s == null || s.length() == 0) {
            return s; 
        }
        
        String[] arr = s.split(" "); 
        
        StringBuffer sb = new StringBuffer(); 
        for (int i = arr.length - 1; i >= 0; --i) {
            if (arr[i].equals("")) {
                continue; 
            }
            sb.append(arr[i] + " "); 
        }
        
        if (sb.length() == 0) {
            return ""; 
        }
        sb.deleteCharAt(sb.length() - 1); 
        return sb.toString(); 
    }
}

File: src/CPP/src/leetcode/ReverseWordsInAString.cpp
/*
 * [Source] https://leetcode.com/problems/reverse-words-in-a-string/
 * [Difficulty]: Medium
 * [Tag]: String
 */

#include <iostream>

using namespace std;

// [Solution]: Reverse each word, then reverse the whole string
// [Corner Case]: trailing spaces count as one space in the result
class Solution {
};

// [Solution]: split the string by space
/* Java solution
https://github.com/ZhenyingZhu/ClassicAlgorithms/blob/master/src/algorithms/arrandstring/ReverseWordsInAString.java
 */

/* Java solution
public class Solution {
    public String reverseWords(String s) {
        String[] array=s.split(" ");
        StringBuffer result=new StringBuffer();
        if(array.length<1) return result.toString();
        for(int i=array.length-1; i>=0; i--){
            if(!array[i].equals("")){
                result.append(array[i]+" ");
            }
        }
        String answer=result.toString();
        if(answer.equals("")) return answer; 
        return answer.substring(0, answer.length()-1);
    }
}
 */

int main() {
    Solution sol;

    return 0;
}
